### PR TITLE
[Reformer P1] Add loss, return_dict, hidden and attention outputs

### PR DESCRIPTION
**Pull request by tsinghua-zhang**

### CODE REVIEW

**Review by wj-Mcat — COMMENTED** [review]
1. 在 reformer 当中最初是没有基于 dataclass 来构造Output，这部分属于历史遗留问题，所以64-75 行的`*Output`相关的nametuple都要转化成dataclass来构造输出。

File: paddlenlp/transformers/reformer/modeling.py [hunk 1]
```diff
@@ -2929,5 +2995,14 @@ def forward(
...
+            output = (start_logits, end_logits) + reformer_outputs[1:]
+            return ((total_loss,) + output) if total_loss is not None else output
+
+        return QuestionAnsweringModelOutput(
+            loss=total_loss,
+            start_logits=start_logits,
+            end_logits=end_logits,
+            hidden_states=reformer_outputs.hidden_states,
+            attentions=reformer_outputs.attentions,
+        )
```
Comment (wj-Mcat) [inline paddlenlp/transformers/reformer/modeling.py:2995]: 在这块的模型组网这边有一些问题：

1. dropout = (None,) 这行代码需要删掉，估计是写的时候就没有注意。
2. 另外，transformers 那边是没有`dropout`这一层的，所以咱们这边也是需要同步删掉的。
3. 现在有了`PretrainedConfig`模块，所以类似于 `self.reformer.config["hidden_dropout_prob"]` 的代码都应该调整成为：`self.reformer.config.hidden_dropout_prob`。BTW：我搜索了下大概有 20 处左右出现这样的问题。

File: paddlenlp/transformers/reformer/modeling.py [hunk 2]
```diff
@@ -2908,6 +2973,7 @@ def forward(
             use_cache=False,  # no causal mask
             output_hidden_states=output_hidden_states,
             output_attentions=output_attentions,
+            return_dict=return_dict,
```
Comment (wj-Mcat) [inline paddlenlp/transformers/reformer/modeling.py:2963]: 另外，在所有的`ReformerForXX`模型中，`return_dict`的初始化都应该在模型的 forward 中完成，不然当return_dict为 None 的时候，没办法通过self.config.return_dict来初始化，具体可参考：

https://github.com/PaddlePaddle/PaddleNLP/blob/c35b670699a114b39ff4f78ab72dd1dc343a56a1/paddlenlp/transformers/bert/modeling.py#L705

Comment (wj-Mcat) [reply paddlenlp/transformers/reformer/modeling.py:2963]: 其它模型类似的地方都应该调整。

File: paddlenlp/transformers/reformer/modeling.py (hunk 1 above)
Comment (wj-Mcat) [reply paddlenlp/transformers/reformer/modeling.py:2995]: 此外，如果是`ReformerForXXX`模型，可直接`self.config`来获取`ReformerConfig`对象，可以不用`self.reformer.config`。

Comment (wj-Mcat) [reply paddlenlp/transformers/reformer/modeling.py:2995]: 你现在给部分函数的参数添加了type hint, 可是其他的参数都没有添加，所以我建议其他的都添加上把：目前可只修改：forward 函数就行。

例如：

https://github.com/PaddlePaddle/PaddleNLP/blob/9ef80869adbd2b311720e31a84c91365833e6035/paddlenlp/transformers/reformer/modeling.py#L2877-L2888

这里：input_ids 参数都添加上。

> 如果这里有input_embed参数 (P2 的功能)，input_ids 可以为 None，可是目前还支持P1 阶段，严谨上来讲是不应该为 None，不过以后注意了，现在就先这样。

File: tests/transformers/reformer/test_modeling.py [hunk 3]
```diff
@@ -275,8 +278,12 @@ def create_and_check_reformer_model_with_attn_mask(
...
-        output_padded = model(input_ids_padded, attention_mask=attn_mask)[0][:, :half_seq_len]
-        output_padded_rolled = model(input_ids_roll, attention_mask=attn_mask_roll)[0][:, roll : half_seq_len + roll]
+        output_padded = model(input_ids_padded, attention_mask=attn_mask, return_dict=self.parent.return_dict)[0][
+            :, :half_seq_len
+        ]
+        output_padded_rolled = model(
+            input_ids_roll, attention_mask=attn_mask_roll, return_dict=self.parent.return_dict
+        )[0][:, roll : half_seq_len + roll]
 
         self.parent.assertTrue(paddle.allclose(output_padded, output_padded_rolled, atol=1e-3))
```
Comment (wj-Mcat) [inline tests/transformers/reformer/test_modeling.py]: 这里的 atol=1e-3 还是有点小了，咱们模型单测精度统一使用1e-4 。

**Review by wj-Mcat — DISMISSED** [review]
整体上的实现结构我觉得很好，就是在小细节上面需要调整。

File: paddlenlp/transformers/reformer/modeling.py [hunk 4]
```diff
@@ -2004,9 +2050,7 @@ def _init_weights(self, layer):
             layer.weight.set_value(
                 paddle.tensor.normal(
                     mean=0.0,
-                    std=self.axial_norm_std
-                    if hasattr(self, "axial_norm_std")
-                    else self.reformer.config["axial_norm_std"],
+                    std=self.axial_norm_std if hasattr(self, "axial_norm_std") else self.config.axial_norm_std,
```
Comment (sijunhe) [inline paddlenlp/transformers/reformer/modeling.py]: 直接self.config.axial_norm_std 就可以了

File: paddlenlp/transformers/reformer/modeling.py [hunk 5]
```diff
@@ -1994,7 +2040,7 @@ def _init_weights(self, layer):
                     mean=0.0,
                     std=self.initializer_range
                     if hasattr(self, "initializer_range")
-                    else self.reformer.config["initializer_range"],
+                    else self.config.initializer_range,
```
Comment (sijunhe) [inline paddlenlp/transformers/reformer/modeling.py]: 直接self.config. initializer_range

File: paddlenlp/transformers/reformer/modeling.py [hunk 6]
```diff
@@ -1981,9 +2029,7 @@ def _init_weights(self, layer):
                 weight.set_value(
                     paddle.tensor.normal(
                         mean=0.0,
-                        std=self.axial_norm_std
-                        if hasattr(self, "axial_norm_std")
-                        else self.reformer.config["axial_norm_std"],
+                        std=self.axial_norm_std if hasattr(self, "axial_norm_std") else self.config.axial_norm_std,
```
Comment (sijunhe) [inline paddlenlp/transformers/reformer/modeling.py]: 直接self.config.axial_norm_std

File: paddlenlp/transformers/reformer/modeling.py [hunk 7]
```diff
@@ -2679,27 +2730,23 @@ def __init__(self, config: ReformerConfig):
...
         self.num_classes = config.num_classes
-        dropout = None
-        classifier_dropout = dropout if dropout is not None else self.reformer.config["hidden_dropout_prob"]
-        self.classifier = ReformerClassificationHead(
-            hidden_size=self.reformer.config["hidden_size"],
-            classifier_dropout=classifier_dropout,
-            num_classes=config.num_classes,
-        )
-        if self.reformer.config["is_decoder"] is True:
+        self.config = config
```
Comment (sijunhe) [inline paddlenlp/transformers/reformer/modeling.py]: 这行应该不用，self.config是基类自带的

File: paddlenlp/transformers/reformer/modeling.py [hunk 8]
```diff
@@ -2679,27 +2730,23 @@ def __init__(self, config: ReformerConfig):
...
+        self.classifier = ReformerClassificationHead(config)
+        if self.config.is_decoder is True:
```
Comment (sijunhe) [inline paddlenlp/transformers/reformer/modeling.py]:
```suggestion
        if self.config.is_decoder:
```

File: paddlenlp/transformers/reformer/modeling.py [hunk 9]
```diff
@@ -2805,25 +2865,24 @@ class ReformerForQuestionAnswering(ReformerPretrainedModel):
...
     def __init__(self, config: ReformerConfig):
         super().__init__(config)
-        dropout = (None,)
         self.reformer = ReformerModel(config)
-        self.dropout = nn.Dropout(dropout if dropout is not None else self.reformer.config["hidden_dropout_prob"])
         # 2 * hidden_size because we use reversible residual layers
-        self.qa_outputs = nn.Linear(2 * self.reformer.config["hidden_size"], 2)
-        if self.reformer.config["is_decoder"] is True:
+        self.qa_outputs = nn.Linear(2 * self.config.hidden_size, 2)
+        if self.config.is_decoder is True:
```
Comment (sijunhe) [inline paddlenlp/transformers/reformer/modeling.py]:
```suggestion
        if self.config.is_decoder:
```

File: paddlenlp/transformers/reformer/modeling.py [hunk 10]
```diff
@@ -1911,6 +1909,56 @@ def forward(self, hidden_states):
         return hidden_states
 
 
+@dataclass
```
Comment (sijunhe) [inline paddlenlp/transformers/reformer/modeling.py:1910]: 这里所有的dataclass 都多了一行空格，可以删除

**Review by sijunhe — COMMENTED** [review]
挺好，修一下小东西就可以合入了

**Review by sijunhe — APPROVED** [review]
lgtm

**Review by sijunhe — APPROVED** [review]
great job!

**Comment by wj-Mcat** [comment]
@sijunhe  thanks for reviewing.
